File: 2026-08/src/day_22/prms_42898/Solution.py
MOD = 1_000_000_007


def solution(m, n, puddles):
    chk = [[False] * (1 + m) for _ in range(1 + n)]
    for x, y in puddles:
        chk[y][x] = True

    dp = [[0] * (1 + m) for _ in range(1 + n)]
    dp[0][1] = 1

    for i in range(1, 1 + n):
        for j in range(1, 1 + m):
            if chk[i][j]:
                continue
            dp[i][j] = (dp[i - 1][j] + dp[i][j - 1]) % MOD

    return dp[n][m]

File: 2026-08/src/day_22/prms_42898/Solution.java
package day_22.prms_42898;

class Solution {

    static final int MOD = 1_000_000_007;

    public int solution(int m, int n, int[][] puddles) {
        boolean[][] chk = new boolean[1 + n][1 + m];
        for (int[] p : puddles) {
            chk[p[1]][p[0]] = true;
        }

        int[][] dp = new int[1 + n][1 + m];
        dp[0][1] = 1;

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (chk[i][j]) continue;
                dp[i][j] = (dp[i - 1][j] + dp[i][j - 1]) % MOD;
            }
        }

        return dp[n][m];
    }
}
